### ISSUE TITLE
NameOf Analyzer is too noisy

### ISSUE THREAD

**Issue by paul1956** [code-cracker/code-cracker#518]
Is there any way to limit the Use NameOf analyzer to only issue warnings when the variable is something I could rename. I am getting suggestions on "String", "Format", ("Address", "CompanyName", "State", "City" because I reference Outlook), "MS" in a list of states and all kinds of other places where it makes no sense. I would be nice if it only offer the fix for variable in my project and not all the libraries I might include.

**Comment by carloscds** [comment]
@paul1956 Can you send a sample code about this ?

**Comment by paul1956** [comment]
```
Imports Microsoft.Office.Interop

<TestClass()> Public Class UnitTest5
    Private OutlookApp As Outlook._Application
    Private OutlookNS As Outlook._NameSpace

    <TestMethod()> Public Sub TestMethod1()
        Dim A As String = "Categories"  ' From Outlook
        Dim B As String = "Outlook"  ' From Outlook
        Dim C As String = "Format"  ' Strings.Format
        Dim D As String = "MS"  ' From Namespace MS
    End Sub
End Class
```

**Comment by paul1956** [comment]
New example has less unrelated issues.

**Comment by paul1956** [comment]
Replacing GetProgramElementNameThatMatchesStringLiteral with the code below seems to correct all the issues I am seeing. I go from 87 Warnings to 0 in a short Outlook Extension. I would be nicer if the namespaces to exclude didn't have to be hard coded, but for now this works. Sorry this is not a pull request I don't understand Git enough to even know where to begin. I have also entered a suggestion to the Roslyn team to look at enhancing model.LookupSymbols so this kind of code is in one place.
```
        Private Shared Function GetProgramElementNameThatMatchesStringLiteral(stringLiteral As LiteralExpressionSyntax, model As SemanticModel) As String
            Dim programElementName As String = GetParameterNameThatMatchesStringLiteral(stringLiteral)
            If Not Found(programElementName) Then
                Dim literalValueText As String = stringLiteral.Token.ValueText
                Dim symbols As ImmutableArray(Of ISymbol) = model.LookupSymbols(stringLiteral.Token.SpanStart, Nothing, literalValueText)

                Dim symbol As ISymbol = symbols.FirstOrDefault()
                If symbol Is Nothing Then
                    Return String.Empty
                End If
                If symbol.Name = "MS" Then
                    Return String.Empty
                End If
                If symbol.ToDisplayString.StartsWithAny("Microsoft.", "System.") Then
                    Return String.Empty
                End If
                If symbol.ContainingAssembly.Name.StartsWithAny("Microsoft.", "System.") Then
                    Return String.Empty
                End If
                If symbol?.Kind = SymbolKind.Local Then
                    ' Only register if local variable is declared before it is used.
                    ' Don't recommend if variable is declared after string literal is used.
                    Dim symbolSpan As TextSpan = symbol.Locations.Min(Function(i) i.SourceSpan)
                    If symbolSpan.CompareTo(stringLiteral.Token.Span) > 0 Then
                        Return String.Empty
                    End If
                End If
                programElementName = symbol?.ToDisplayParts().
                        Where(AddressOf IncludeOnlyPartsThatAreName).
                        LastOrDefault(Function(displayPart) displayPart.ToString() = literalValueText).
                        ToString()
            End If
            Return programElementName
        End Function
```
You might also need to add StartsWithAny to your Extensions Module
```
    <Extension>
    Public Function StartsWithAny(text As String, comparisonType As StringComparison, ParamArray values As String()) As Boolean
        For Each value As String In values
            If text.StartsWith(value, comparisonType) Then
                Return True
            End If
        Next
        Return False
    End Function
    <Extension>
    Public Function StartsWithAny(ByVal text As String, ParamArray values() As String) As Boolean
        Return text.StartsWithAny(StringComparison.CurrentCulture, values)
    End Function
```

**Comment by giggio** [comment]
I think it makes sense. We have a related issue discussing exactly how far this analyzer should go. We haven't decided yet.

**Comment by paul1956** [comment]
I Tried my fix on another application that uses USPS API and got 100's of incorrect NameOf warnings. As soon as I added that filter it highlighted a few real issues. For this to be really useful there needs to be a way to list Dll's and Namespaces to ignore and I have no idea how to add such a Dialog Box to an Analyzer.

**Comment by JimiC** [comment]
To add on the discussion, this rule should be ignored in auto-generated files like the GlobalSuppression.

**Comment by akamud** [comment]
I have an example in a Code Cracker fix provider. I was trying to write a VB code fix provider that needed this code:

```VB.NET
Private Shared ReadOnly AnyName As SimpleNameSyntax = TryCast(SyntaxFactory.ParseName("Any"), SimpleNameSyntax)
```

The NameOf analyzer won't let me use the `"Any"` string because there is a method `Any` in `Microsoft.CodeAnalysis.VisualBasicExtensions`.  
Maybe it should ignore extension methods?

**Comment by paul1956** [comment]
@akamud The purpose of NameOf is to allow renaming of variables that are passed as a string, usually for error messages. But you don't want to globally rename things in libraries (which would fail anyway) so you don't need to use NameOf and the analyzer should not offer it.

**Comment by akamud** [comment]
Exactly @paul1956, I had no option in this case.

**Comment by giggio** [comment]
I think @paul1956 nailed it in his last comment. It makes sense to me that `nameof` is something we use to reference our own symbols, so if we rename them by accident we break the build, and it also gets picked up by renaming refactorings. It makes no sense to have the analyzer raise a diagnostic in other places. It should not create a diagnostic. Or it should be optional.
We could make the current analyzer to split the analysis into 2 diagnostic ids, one for the user's code, and another for imported symbols, which would come disabled. Or maybe we simply ignore it.
I am not really sure how to fix this, I don't know if Roslyn let's us differentiate between an imported symbol and a internal solution symbol. Let's hope it does.

**Comment by paul1956** [comment]
I did a quick fix and have it ignore reference that start with "Microsoft.", "MS." and "System.", I am also playing with a search for all External references and a way for users to add their own list. Below is a piece of the working code. I have been running into lots of issue getting everything to work. It is rumored that if any file in the solution is changed all the Analyzers will run again. So far I have not found that to be true which is making a user entered list difficult without exiting VS between changing the list.
```
    Private IgnoreList As New List(Of String) With {"Microsoft."."System.","MS."}


            Dim workspace As VisualStudioWorkspace = PureVisualBasicPackage.MyComponentModel.GetService(Of VisualStudioWorkspace)()
            Dim SolutionPath As String = workspace.CurrentSolution.FilePath
            For Each project As Project In workspace.CurrentSolution.Projects
                For Each reference As MetadataReference In project.MetadataReferences
                    'If MetadataReferenceList.Contains(reference.Display) Then
                    '    Continue For
                    'End If
                    MetadataReferenceList.Add(reference.Display)
                    Dim SplitFileNameReference() As String = Path.GetFileName(reference.Display).Split(CChar("."))
                    If SplitFileNameReference.Length <= 2 OrElse IgnoreList.Contains(SplitFileNameReference(0)) Then
                        Continue For
                    End If
                    IgnoreList.Add(SplitFileNameReference(0))
                Next
```

**Comment by giggio** [comment]
I fixed it on #531. There is a way to find the source. I am waiting for @carloscds or @ElemarJR to review my PR.

**Comment by giggio** [comment]
@paul1956 It is now on myget with version 1.0.0-rc3-0849 (and laters). It will be on Nuget on rc4 or 1.0.0.
